src/ast/desugar.rs

```diff
@@ -375,11 +383,77 @@ pub(crate) fn desugar_command(
             if name == "".into() {
                 name = rule.to_string().replace('\"', "'").into();
             }
-            vec![NCommand::NormRule {
-                ruleset,
-                name,
-                rule: flatten_rule(rule, desugar),
-            }]
+
+            if seminaive {
+                let mut new_rule = rule.clone();
+                // only add new rule when there is Call in body to avoid adding same rule.
+                let mut add_new_rule = false;
+
+                for head_slice in new_rule.head.iter_mut() {
+                    match head_slice {
+                        Action::Set(_, _, value) => {
+                            // if the right hand side is a function call,
+                            // move it to body so seminaive fires
+                            match value {
```

Comment:
This can probably be simplified to `if matches!(value, Expr::Call(_, _)) {`. Also, I think `value` here does not necessarily have to be a `Call`? For example,

```
(rule (...)
    ((let x (f 1))
     (set y x)))
```
In the future, we can potentially do some reachability/type analysis to find out what needs to be moved to the body and what needn't. For now we can keep things simple (probably good to leave a comment here mentioning potential optimizations for future reference).